File: php/bob/bob.php
<?php

class Bob {
  public function respondTo($input) {
    if ($this->forcefullQuestioning($input)) {
      return "Calm down, I know what I'm doing!";
    } elseif ($this->yelling($input)) {
      return "Whoa, chill out!";
    } elseif ($this->questioning($input)) {
      return "Sure.";
    } elseif ($this->silence($input)) {
      return "Fine. Be that way!";
    }
    return "Whatever.";
  }

  private function silence($input) {
    return empty(trim(preg_replace('/\pZ+/u', '', $input)));
  }

  private function yelling($input) {
    return $this->containsLetters($input) && $input === mb_strtoupper($input);
  }

  private function questioning($input) {
    return $this->endsWith(trim($input), '?');
  }

  private function forcefullQuestioning($input) {
    return $this->yelling($input) && $this->questioning($input);
  }

  private function endsWith($input, $expected) {
    return mb_substr($input, -mb_strlen($expected)) === $expected;
  }

  function containsLetters($input) {
    return preg_match( '/[a-zA-Z]/', $input );
}
}

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    cond do
      silence?(input) -> "Fine. Be that way!"
      shouting?(input) && questioning?(input) -> "Calm down, I know what I'm doing!"
      questioning?(input) -> "Sure."
      shouting?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  defp shouting?(input) do
    upped = String.upcase(input)
    downed = String.downcase(input)
    input == upped && upped != downed
  end

  defp questioning?(input) do
    String.ends_with?(input, "?")
  end

  defp silence?(input) do
    String.trim(input) == ""
  end
end
